lib/recursive-methods.rb

```diff
@@ -1,49 +1,119 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0 
+        raise ArgumentError,'only input positivie integer'
+    end 
+    if (n == 1 || n == 0 )
+        return 1
+    else 
+        return n * factorial(n-1)
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: 0(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <=1 
+        return s 
+    end 
+    reverseword = ""
+    i = s.length 
+    while i > 0 
+        reverseword += s[i-1]
+        i -= 1
+    end 
+    return reverseword
+
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
```

Comment:
This works, but it's not in-place.  This is because you're making a new array with each recursive call.

The below does the job without creating a new string with each call. 

```ruby
def reverse_inplace(s, first = 0, last = s.length -1)
  return s if first < last
  s[first], s[last] = s[last], s[first]
  return reverse_inplace(s, first + 1, last  - 1)
end
```